torch/lowrank.py

```diff
@@ -0,0 +1,267 @@
+    :math:`A \approx U diag(S) V^T`. In case :math:`M` is given, then
+    SVD is computed for the matrix :math:`A - M`.
+
+    .. note:: The implementation is based on the Algorithm 5.1 from
+              Halko et al, 2009.
+
+    .. note:: To obtain repeatable results, reset the seed for the
+              pseudorandom number generator
+
+    .. note:: The input is assumed to be a low-rank matrix.
+
+    .. note:: In general, use the full-rank SVD implementation
+              ``torch.svd`` for dense matrices due to its 10-fold
+              higher performance characteristics. The low-rank SVD
+              will be useful for huge sparse matrices that
+              ``torch.svd`` cannot handle.
+
+    Arguments::
+        A (Tensor): the input tensor of size :math:`(*, m, n)`
+
+        q (int, optional): a slightly overestimated rank of A.
+
+        niter (int, optional): the number of subspace iterations to
+                               conduct; niter must be a nonnegative
+                               integer, and defaults to 2
+
+        M (Tensor, optional): the input tensor's mean of size
+                              :math:`(*, 1, n)`.
+
+    References::
+        - Nathan Halko, Per-Gunnar Martinsson, and Joel Tropp, Finding
+          structure with randomness: probabilistic algorithms for
+          constructing approximate matrix decompositions,
+          arXiv:0909.4061 [math.NA; math.PR], 2009 (available at
+          `arXiv <http://arxiv.org/abs/0909.4061>`_).
+
+    """
+    m, n = A.shape[-2:]
+    matmul = _utils.get_matmul(A)
+    if M is None:
+        M_t = None
+    else:
+        matmul2 = _utils.get_matmul(M)
+        M_t = _utils.transpose(M)
+    A_t = _utils.transpose(A)
+
+    # Algorithm 5.1 in Halko et al 2009, slightly modified to reduce
+    # the number conjugate and transpose operations
+    if m < n:
+        # computing the SVD approximation of a transpose in order to
+        # keep B shape minimal
+        Q = get_approximate_basis(A_t, q, niter=niter, M=M_t)
+        Q_c = _utils.conjugate(Q)
+        if M is None:
+            B_t = matmul(A, Q_c)
+        else:
+            B_t = matmul(A, Q_c) - matmul2(M, Q_c)
+        U, S, V = torch.svd(B_t)
+        V = Q.matmul(V)
+    else:
+        Q = get_approximate_basis(A, q, niter=niter, M=M)
+        Q_c = _utils.conjugate(Q)
+        if M is None:
+            B = matmul(A_t, Q_c)
+        else:
+            B = matmul(A_t, Q_c) - matmul2(M_t, Q_c)
+        U, S, V = torch.svd(_utils.transpose(B))
+        U = Q.matmul(U)
+
+    return U, S, V
+
+
+def pca(A, q=None, center=True, niter=2):
+    r"""Performs Principal Component Analysis (PCA) on a low-rank matrix,
+    batches of such matrices, or sparse matrix.
+
+    This function returns a namedtuple ``(U, S, V)`` which is the
+    nearly optimal approximation of a singular value decomposition of
+    a centered matrix :attr:`A` such that :math:`A = U diag(S) V^T`.
```

Comment:
I'm wondering if users would be more familiar with PCA returning the principal directions `V`, the principal components `US`, and the eigenvalue of the covariance matrix `\lambda_i = s_i^2/(n-1)`, when centered. I like that PCA and SVD return similar output, so I would simply add a note highlighting this in the documentation. Thoughts?

Reply:
I think it is a good idea. Added the following note:
```
    .. note:: The relation of ``(U, S, V)`` to PCA is as follows:

                - ``A`` is the data matrix with ``m`` samples and
                  ``n`` features

                - the ``V`` columns represent the principal directions

                - ``S**2/(m-1)`` are the eigenvalues of covariance
                  matrix ``A^T A/(m-1)``

                - ``A V[:,:k]`` projects data to the first k principal
                  components
```

Reply:
> ```
>                 - ``S**2/(m-1)`` are the eigenvalues of covariance
>                   matrix ``A^T A/(m-1)``
> ```

nit: `A^T A / (m-1)` is only the covariance matrix when `A` itself has zero mean. How about
```
``S**2/(m-1)`` contains the eigenvalues of ``A^T A/(m-1), the covariance of ``A`` when ``A`` is centered, or ``center=False`` is provided.
```

By the way, I'm assuming `center=True` means that the `A` is centered, and, when `center=False`, the mean of `A` is removed. Is that correct?

Reply:
No, `center=True` means that the `A` will be centered: "center" is used as a verb here.
To avoid confusion, consider other options:
1. use `centered=False` to mean that `A` needs to be centered
2. use `raw=True` to mean that `A` needs to be centered (taken from https://github.com/facebook/fbpca/blob/f32210ea90f14845cadb0214a3fdd0246fa50f70/fbpca.py#L1549)
3. other ideas?
I'd prefer option 1 or current behavior. What do you think?

Reply:
Let's keep `center=True` and just rephrase the description to
```
        center (bool, optional): if True, center the input tensor.
```

Reply:
Fixed.